### ISSUE TITLE
Allocation of `Kokkos::View` given a `space` instance and `Cuda UVM`

### ISSUE THREAD

**Issue by romintomasetti** [kokkos/kokkos#6005]
**Describe the bug**

Take the following example (with `Cuda UVM` enabled):
```c++
// Get a Cuda stream different from the default stream
const auto space = Kokkos::Experimental::partition_space(Kokkos::Cuda{}, 1);
// Build a view, giving the stream
Kokkos::View<double*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>> view(
    Kokkos::view_alloc(space, "my label"),
    size
);
```

In the general case, it seems it is safe to access the view from host at this point because of the following lines doing a `fence` on the default stream, and the default stream being by default (`legacy mode`) "device-wide synchronizing":
https://github.com/kokkos/kokkos/blob/aa1f48f3172069f212ed2c97b74b786999b26f17/core/src/Kokkos_View.hpp#L1420-L1442

However, if one enables the `per-thread` default stream mode, I think the aforementioned lines are still fencing the default stream, but now the default stream fence does not affect the custom stream used in the allocation. It means that accessing the view from host can be hazardous.

A potential fix could be to determine if a `space` instance was given, and fence on it instead of fencing on the default stream:
```c++
if constexpr (alloc_prop::has_execution_space) {
    const auto& exec_space = Impl::get_property<Impl::ExecutionSpaceTag>(arg_prop);
    exec_space.fence(...);
} else {
    typename traits::device_type::memory_space::execution_space().fence(
          "Kokkos::View<...>::View: fence before allocating UVM");
}
```

Also tagging @maartenarnst as we work on that together.

**Comment by PhilMiller** [comment]
I think your proposed fix does the right thing when passing an explicit execution space instance.

The fence on the default instance may actually not be appropriate anymore if we actually rely on device-wide synchronization, because the default-constructed `Kokkos::Cuda` instance does not use the 'default' CUDA stream, but rather a separate one we create internally, precisely to avoid the implicit global synchronization semantics. When we or users truly want device-wide synchronization, rather than synchronization of the operations on the particular stream, that requires calling `Kokkos::fence()` rather than `exec_space.fence()`.

**Comment by masterleinad** [comment]
@romintomasetti Would you mind creating a pull request for this? Note that `prop_copy` already contains the appropriate execution space instance that should be used to fence.

**Comment by PhilMiller** [comment]
So the simplified implementation would just be `Impl::get_property<Impl::ExecutionSpaceTag>(prop_copy).fence();`, but we still need to figure whether it's still supposed to be a global fence in the not-provided case.

**Comment by masterleinad** [comment]
I really don't see a reason why this should be a global fence.

**Comment by romintomasetti** [comment]
@masterleinad @PhilMiller

I created #6008 for this. Tell me if the PR style is OK for you ;)

**Comment by dalg24** [comment]
The fix for this issue was merged in develop and will be in release 4.1
